rdrf/rdrf/schema/schema.py

```diff
@@ -0,0 +1,224 @@
+                cde_value = cde['value'] if type(cde['value']) is list else [(cde['value'])]
+                values.append(ClinicalDataTypeFlat(form['name'], section['code'], cde['code'], cde_value))
+
+        for entry in clinical_data:
+            if 'forms' in entry.data:
+                for form in entry.data['forms']:
+                    for section in form['sections']:
+                        for cde in section['cdes']:
+                            if 'allow_multiple' in section and section['allow_multiple'] is True:
+                                logger.info("this is a multi section")
+                                logger.info(cde)
+                                for cde_entry in cde:
+                                    add_value(form, section, cde_entry, cde_keys)
+                            else:
+                                add_value(form, section, cde, cde_keys)
+        return values
+
+    # def resolve_clinical_data(self, info, cde_keys=[]):
+    #
+    #     # TODO only resolve clinical data with matching cde_keys
+    #     clinical_data = ClinicalData.objects.filter(django_id=self.id, django_model='Patient', collection="cdes")
+    #
+    #     context_form_ids = clinical_data.values_list("context_id", flat=True)
+    #     context_form_group_names = RDRFContext.objects.filter(pk__in=list(context_form_ids)).values("pk",
+    #                                                                                                 "context_form_group__name")
+    #     cfg_lookup = {cfg["pk"]: cfg["context_form_group__name"] for cfg in context_form_group_names}
+    #
+    #     values = {}
+    #
+    #     for clinical_datum in clinical_data:
+    #         clinical_data_forms = [{
+    #             "name": form["name"],
+    #             "sections": form["sections"],
+    #             "timestamp": clinical_datum.data["timestamp"]
+    #         } for form in clinical_datum.data["forms"]]
+    #
+    #         values.setdefault(cfg_lookup[clinical_datum.context_id], []).extend(clinical_data_forms)
+    #
+    #     return {"context_form_groups": [{"name": key, "forms": value} for key, value in values.items()]}
+
+    def resolve_clinical_data(self, info, cde_keys=[]):
+
+        def is_multisection(section):
+            return 'allow_multiple' in section and section['allow_multiple'] is True
+
+        clinical_data = ClinicalData.objects.filter(django_id=self.id, django_model='Patient', collection="cdes")
+
+        context_form_ids = clinical_data.values_list("context_id", flat=True)
+        context_form_group_names = RDRFContext.objects.filter(pk__in=list(context_form_ids)).values("pk",
+                                                                                                    "context_form_group__name")
+        cfg_lookup = {cfg["pk"]: cfg["context_form_group__name"] for cfg in context_form_group_names}
+
+        values = {}
+        for datum in clinical_data:
```

Comment:
@ScriptSmith This is the ugly multi nested loop I was talking about.

Reply:
I can rewrite it like this:

```python
    def resolve_clinical_data(self, info, cde_keys=[]):
        clinical_data = ClinicalData.objects.filter(django_id=self.id, django_model='Patient', collection="cdes")

        context_form_ids = clinical_data.values_list("context_id", flat=True)
        context_form_group_names = RDRFContext.objects.filter(pk__in=list(context_form_ids)).values("pk",
                                                                                                    "context_form_group__name")
        cfg_lookup = {cfg["pk"]: cfg["context_form_group__name"] for cfg in context_form_group_names}

        values = {}
        for datum in clinical_data:
            values.setdefault(cfg_lookup[datum.context_id], []).extend(list(filter(lambda form: form['sections'], [{
                'name': form['name'],
                'timestamp': datum.data['timestamp'],
                'sections': list(filter(lambda section: section['cdes'], [{
                    'code': section['code'],
                    'allow_multiple': section['allow_multiple'],
                    'cdes': list(filter(None, [
                        [c for c in cde if mongo_key(form['name'], section['code'], c['code']) in cde_keys]
                        for cde in section['cdes']
                    ] if 'allow_multiple' in section and section['allow_multiple'] else [
                        cde for cde in section['cdes'] if
                        mongo_key(form['name'], section['code'], cde['code']) in cde_keys
                    ]))
                } for section in form['sections']]))
            } for form in datum.data['forms']])))

        return {"context_form_groups": [{"name": key, "forms": value} for key, value in values.items() if value]}
```

That's a first pass, and there are a few ways to simplify it even further. However, I'd argue it's probably worse in terms of readability and debugability than the current implementation.

I think there are some techniques you could use in python's `itertools` and `collections` modules (eg. ChainMap) to build it in a different way. I'll sleep on it and see what I think tomorrow